File: src/packages/sdk/src/otel.ts
/**
 * OpenTelemetry Export for WorkflowTracer
 *
 * Converts WorkflowTracer spans, decisions, and costs into
 * OpenTelemetry-compatible span data for export to any OTEL collector.
 *
 * Usage:
 *   import { OTelExporter } from "@evalgate/sdk/otel";
 *
 *   const exporter = new OTelExporter({ endpoint: "http://localhost:4318" });
 *   const tracer = new WorkflowTracer(client, { debug: true });
 *   // ... run workflow ...
 *   await exporter.exportFromTracer(tracer);
 */

import type {
	AgentHandoff,
	CostRecord,
	RecordDecisionParams,
	WorkflowTracer,
} from "./workflows";

/**
 * OTEL-compatible span representation
 * Follows the OpenTelemetry Trace specification
 */
export interface OTelSpan {
	traceId: string;
	spanId: string;
	parentSpanId?: string;
	name: string;
	/** OTLP SpanKind: 0=UNSPECIFIED, 1=INTERNAL, 2=SERVER, 3=CLIENT, 4=PRODUCER, 5=CONSUMER */
	kind: 0 | 1 | 2 | 3 | 4 | 5;
	startTimeUnixNano: string;
	endTimeUnixNano: string;
	attributes: OTelAttribute[];
	/** OTLP StatusCode: 0=STATUS_CODE_UNSET, 1=STATUS_CODE_OK, 2=STATUS_CODE_ERROR */
	status: { code: 0 | 1 | 2; message?: string };
	events: OTelEvent[];
}

export interface OTelAttribute {
	key: string;
	value: {
		stringValue?: string;
		intValue?: string;
		doubleValue?: number;
		boolValue?: boolean;
	};
}

export interface OTelEvent {
	name: string;
	timeUnixNano: string;
	attributes: OTelAttribute[];
}

/**
 * OTEL export payload (OTLP JSON format)
 */
export interface OTelExportPayload {
	resourceSpans: Array<{
		resource: {
			attributes: OTelAttribute[];
		};
		scopeSpans: Array<{
			scope: { name: string; version: string };
			spans: OTelSpan[];
		}>;
	}>;
}

export interface OTelExporterOptions {
	/** OTEL collector endpoint (default: http://localhost:4318/v1/traces) */
	endpoint?: string;
	/** Service name for resource attributes */
	serviceName?: string;
	/** Additional resource attributes */
	resourceAttributes?: Record<string, string>;
	/** SDK version */
	sdkVersion?: string;
	/** Headers for the export request */
	headers?: Record<string, string>;
}

/**
 * Generate a random 16-byte hex trace ID
 */
function generateTraceId(): string {
	const bytes = new Uint8Array(16);
	for (let i = 0; i < 16; i++) {
		bytes[i] = Math.floor(Math.random() * 256);
	}
	return Array.from(bytes)
		.map((b) => b.toString(16).padStart(2, "0"))
		.join("");
}

/**
 * Generate a random 8-byte hex span ID
 */
function generateSpanId(): string {
	const bytes = new Uint8Array(8);
	for (let i = 0; i < 8; i++) {
		bytes[i] = Math.floor(Math.random() * 256);
	}
	return Array.from(bytes)
		.map((b) => b.toString(16).padStart(2, "0"))
		.join("");
}

/**
 * Convert milliseconds to nanosecond string
 */
function msToNano(ms: number): string {
	return `${BigInt(ms) * BigInt(1_000_000)}`;
}

/**
 * Create an OTEL attribute
 */
function attr(key: string, value: string | number | boolean): OTelAttribute {
	if (typeof value === "string") {
		return { key, value: { stringValue: value } };
	}
	if (typeof value === "number") {
		if (Number.isInteger(value)) {
			return { key, value: { intValue: String(value) } };
		}
		return { key, value: { doubleValue: value } };
	}
	return { key, value: { boolValue: value } };
}

/**
 * OpenTelemetry Exporter for EvalGate WorkflowTracer
 */
export class OTelExporter {
	private options: Required<OTelExporterOptions>;

	constructor(options: OTelExporterOptions = {}) {
		this.options = {
			endpoint: options.endpoint ?? "http://localhost:4318/v1/traces",
			serviceName: options.serviceName ?? "evalgate",
			resourceAttributes: options.resourceAttributes ?? {},
			sdkVersion: options.sdkVersion ?? "2.3.0",
			headers: options.headers ?? {},
		};
	}

	/**
	 * Export workflow data from a WorkflowTracer instance
	 */
	exportFromTracer(tracer: WorkflowTracer): OTelExportPayload {
		const workflow = tracer.getCurrentWorkflow();
		const handoffs = tracer.getHandoffs();
		const decisions = tracer.getDecisions();
		const costs = tracer.getCosts();

		const traceId = generateTraceId();
		const rootSpanId = generateSpanId();
		const now = Date.now();

		const spans: OTelSpan[] = [];

		// Root workflow span
		if (workflow) {
			spans.push({
				traceId,
				spanId: rootSpanId,
				name: `workflow.${workflow.name}`,
				kind: 1,
				startTimeUnixNano: msToNano(new Date(workflow.startedAt).getTime()),
				endTimeUnixNano: msToNano(now),
				attributes: [
					attr("evalgate.workflow.name", workflow.name),
					attr("evalgate.workflow.id", workflow.id),
					attr("evalgate.workflow.trace_id", workflow.traceId),
				],
				status: { code: 1 },
				events: [],
			});
		}

		// Decision spans
		for (let i = 0; i < decisions.length; i++) {
			const decision = decisions[i];
			const spanId = generateSpanId();
			spans.push(
				this.decisionToSpan(
					traceId,
					spanId,
					rootSpanId,
					decision,
					now - decisions.length + i,
				),
			);
		}

		// Handoff events
		for (let i = 0; i < handoffs.length; i++) {
			const handoff = handoffs[i];
			const spanId = generateSpanId();
			spans.push(this.handoffToSpan(traceId, spanId, rootSpanId, handoff));
		}

		// Cost spans
		for (let i = 0; i < costs.length; i++) {
			const cost = costs[i];
			const spanId = generateSpanId();
			spans.push(
				this.costToSpan(
					traceId,
					spanId,
					rootSpanId,
					cost,
					now - costs.length + i,
				),
			);
		}

		return this.buildPayload(spans);
	}

	/**
	 * Export a run result as OTEL spans
	 */
	exportRunResult(runResult: {
		runId: string;
		metadata: {
			startedAt: number;
			completedAt: number;
			duration: number;
			mode: string;
		};
		results: Array<{
			specId: string;
			name: string;
			filePath: string;
			result: {
				status: string;
				score?: number;
				duration: number;
				error?: string;
			};
		}>;
		summary: { passed: number; failed: number; passRate: number };
	}): OTelExportPayload {
		const traceId = generateTraceId();
		const rootSpanId = generateSpanId();
		const spans: OTelSpan[] = [];

		// Root run span
		spans.push({
			traceId,
			spanId: rootSpanId,
			name: `evalgate.run.${runResult.runId}`,
			kind: 1,
			startTimeUnixNano: msToNano(runResult.metadata.startedAt),
			endTimeUnixNano: msToNano(runResult.metadata.completedAt),
			attributes: [
				attr("evalgate.run.id", runResult.runId),
				attr("evalgate.run.mode", runResult.metadata.mode),
				attr("evalgate.run.duration_ms", runResult.metadata.duration),
				attr("evalgate.run.pass_rate", runResult.summary.passRate),
				attr("evalgate.run.passed", runResult.summary.passed),
				attr("evalgate.run.failed", runResult.summary.failed),
			],
			status: {
				code: runResult.summary.failed > 0 ? 2 : 1,
			},
			events: [],
		});

		// Per-spec child spans
		let offset = 0;
		for (const spec of runResult.results) {
			const spanId = generateSpanId();
			const specStart = runResult.metadata.startedAt + offset;
			const specEnd = specStart + spec.result.duration;
			offset += spec.result.duration;

			const attributes: OTelAttribute[] = [
				attr("evalgate.spec.id", spec.specId),
				attr("evalgate.spec.name", spec.name),
				attr("evalgate.spec.file", spec.filePath),
				attr("evalgate.spec.status", spec.result.status),
				attr("evalgate.spec.duration_ms", spec.result.duration),
			];

			if (spec.result.score !== undefined) {
				attributes.push(attr("evalgate.spec.score", spec.result.score));
			}

			spans.push({
				traceId,
				spanId,
				parentSpanId: rootSpanId,
				name: `evalgate.spec.${spec.name}`,
				kind: 1,
				startTimeUnixNano: msToNano(specStart),
				endTimeUnixNano: msToNano(specEnd),
				attributes,
				status: {
					code: spec.result.status === "passed" ? 1 : 2,
					message: spec.result.error,
				},
				events: [],
			});
		}

		return this.buildPayload(spans);
	}

	/**
	 * Send payload to OTEL collector via HTTP
	 */
	async send(payload: OTelExportPayload): Promise<boolean> {
		try {
			const response = await fetch(this.options.endpoint, {
				method: "POST",
				headers: {
					"Content-Type": "application/json",
					...this.options.headers,
				},
				body: JSON.stringify(payload),
			});
			return response.ok;
		} catch (err) {
			console.warn(
				`[OTelExporter] Failed to send: ${err instanceof Error ? err.message : String(err)}`,
			);
			return false;
		}
	}

	private decisionToSpan(
		traceId: string,
		spanId: string,
		parentSpanId: string,
		decision: RecordDecisionParams,
		timestampMs: number,
	): OTelSpan {
		return {
			traceId,
			spanId,
			parentSpanId,
			name: `decision.${decision.agent}.${decision.chosen}`,
			kind: 1,
			startTimeUnixNano: msToNano(timestampMs),
			endTimeUnixNano: msToNano(timestampMs + 1),
			attributes: [
				attr("evalgate.decision.agent", decision.agent),
				attr("evalgate.decision.type", decision.type),
				attr("evalgate.decision.chosen", decision.chosen),
				attr("evalgate.decision.alternatives", decision.alternatives.length),
				...(decision.confidence !== undefined
					? [attr("evalgate.decision.confidence", decision.confidence)]
					: []),
				...(decision.reasoning
					? [attr("evalgate.decision.reasoning", decision.reasoning)]
					: []),
			],
			status: { code: 1 },
			events: [],
		};
	}

	private handoffToSpan(
		traceId: string,
		spanId: string,
		parentSpanId: string,
		handoff: AgentHandoff,
	): OTelSpan {
		const ts = new Date(handoff.timestamp).getTime();
		return {
			traceId,
			spanId,
			parentSpanId,
			name: `handoff.${handoff.fromAgent ?? "start"}.${handoff.toAgent}`,
			kind: 1,
			startTimeUnixNano: msToNano(ts),
			endTimeUnixNano: msToNano(ts + 1),
			attributes: [
				attr("evalgate.handoff.from", handoff.fromAgent ?? "start"),
				attr("evalgate.handoff.to", handoff.toAgent),
				attr("evalgate.handoff.type", handoff.handoffType),
			],
			status: { code: 1 },
			events: [],
		};
	}

	private costToSpan(
		traceId: string,
		spanId: string,
		parentSpanId: string,
		cost: CostRecord,
		timestampMs: number,
	): OTelSpan {
		return {
			traceId,
			spanId,
			parentSpanId,
			name: `cost.${cost.provider}.${cost.model}`,
			kind: 1,
			startTimeUnixNano: msToNano(timestampMs),
			endTimeUnixNano: msToNano(timestampMs + 1),
			attributes: [
				attr("evalgate.cost.provider", cost.provider),
				attr("evalgate.cost.model", cost.model),
				attr("evalgate.cost.input_tokens", cost.inputTokens),
				attr("evalgate.cost.output_tokens", cost.outputTokens),
				attr("evalgate.cost.total_tokens", cost.totalTokens),
				attr("evalgate.cost.total_usd", cost.totalCost),
			],
			status: { code: 1 },
			events: [],
		};
	}

	private buildPayload(spans: OTelSpan[]): OTelExportPayload {
		const resourceAttrs: OTelAttribute[] = [
			attr("service.name", this.options.serviceName),
			attr("telemetry.sdk.name", "evalgate"),
			attr("telemetry.sdk.version", this.options.sdkVersion),
			attr("telemetry.sdk.language", "nodejs"),
		];

		for (const [key, value] of Object.entries(
			this.options.resourceAttributes,
		)) {
			resourceAttrs.push(attr(key, value));
		}

		return {
			resourceSpans: [
				{
					resource: { attributes: resourceAttrs },
					scopeSpans: [
						{
							scope: {
								name: "evalgate",
								version: this.options.sdkVersion,
							},
							spans,
						},
					],
				},
			],
		};
	}
}

/**
 * Convenience factory
 */
export function createOTelExporter(
	options?: OTelExporterOptions,
): OTelExporter {
	return new OTelExporter(options);
}

File: src/packages/sdk-python/src/evalgate_sdk/otel.py
"""OpenTelemetry exporter for WorkflowTracer spans (T6).

Port of the TypeScript SDK's ``otel.ts``.
Converts workflow tracer data into OTLP-compatible JSON payloads.
"""

from __future__ import annotations

import os
import random
import time
from dataclasses import dataclass, field
from typing import Any

import httpx


def _generate_trace_id() -> str:
    """Generate a 32-hex-char trace ID."""
    return f"{random.getrandbits(128):032x}"


def _generate_span_id() -> str:
    """Generate a 16-hex-char span ID."""
    return f"{random.getrandbits(64):016x}"


def _ms_to_ns(ms: float) -> int:
    """Convert milliseconds to nanoseconds."""
    return int(ms * 1_000_000)


@dataclass
class OTelAttribute:
    key: str
    value: Any

    def to_dict(self) -> dict[str, Any]:
        if isinstance(self.value, bool):
            return {"key": self.key, "value": {"boolValue": self.value}}
        if isinstance(self.value, int):
            return {"key": self.key, "value": {"intValue": str(self.value)}}
        if isinstance(self.value, float):
            return {"key": self.key, "value": {"doubleValue": self.value}}
        return {"key": self.key, "value": {"stringValue": str(self.value)}}


@dataclass
class OTelEvent:
    name: str
    time_unix_nano: int = 0
    attributes: list[OTelAttribute] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "timeUnixNano": str(self.time_unix_nano),
            "attributes": [a.to_dict() for a in self.attributes],
        }


@dataclass
class OTelSpan:
    trace_id: str
    span_id: str
    name: str
    start_time_unix_nano: int
    end_time_unix_nano: int
    parent_span_id: str | None = None
    kind: int = 1  # SPAN_KIND_INTERNAL
    status_code: int = 1  # STATUS_CODE_OK
    status_message: str = ""
    attributes: list[OTelAttribute] = field(default_factory=list)
    events: list[OTelEvent] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "traceId": self.trace_id,
            "spanId": self.span_id,
            "name": self.name,
            "kind": self.kind,
            "startTimeUnixNano": str(self.start_time_unix_nano),
            "endTimeUnixNano": str(self.end_time_unix_nano),
            "status": {"code": self.status_code, "message": self.status_message},
            "attributes": [a.to_dict() for a in self.attributes],
            "events": [e.to_dict() for e in self.events],
        }
        if self.parent_span_id:
            d["parentSpanId"] = self.parent_span_id
        return d


@dataclass
class OTelExportPayload:
    """OTLP JSON export payload."""

    resource_spans: list[dict[str, Any]] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        return {"resourceSpans": self.resource_spans}


@dataclass
class OTelExporterOptions:
    endpoint: str = ""
    service_name: str = "evalgate-sdk"
    headers: dict[str, str] = field(default_factory=dict)
    timeout_ms: int = 10_000


class OTelExporter:
    """Exports evaluation data as OpenTelemetry spans."""

    def __init__(self, options: OTelExporterOptions | None = None) -> None:
        opts = options or OTelExporterOptions()
        self._endpoint = opts.endpoint or os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4318")
        self._service_name = opts.service_name
        self._headers = opts.headers
        self._timeout = opts.timeout_ms / 1000.0

    def export_from_tracer(self, tracer: Any) -> OTelExportPayload:
        """Convert a WorkflowTracer into an OTLP payload.

        *tracer* should expose ``.spans`` (list of dicts with name, start_time,
        end_time, metadata, parent_span_id, etc.).
        """
        trace_id = _generate_trace_id()
        spans: list[OTelSpan] = []

        raw_spans = getattr(tracer, "spans", [])
        for raw in raw_spans:
            span_id = _generate_span_id()
            attrs: list[OTelAttribute] = []

            if isinstance(raw, dict):
                name = raw.get("name", "unknown")
                start_ns = _ms_to_ns(raw.get("start_time", 0))
                end_ns = _ms_to_ns(raw.get("end_time", start_ns))
                parent = raw.get("parent_span_id")
                for k, v in raw.get("metadata", {}).items():
                    attrs.append(OTelAttribute(key=f"evalgate.{k}", value=v))
            else:
                name = getattr(raw, "name", "unknown")
                start_ns = _ms_to_ns(getattr(raw, "start_time", 0))
                end_ns = _ms_to_ns(getattr(raw, "end_time", start_ns))
                parent = getattr(raw, "parent_span_id", None)

            attrs.append(OTelAttribute(key="evalgate.service", value=self._service_name))
            spans.append(
                OTelSpan(
                    trace_id=trace_id,
                    span_id=span_id,
                    name=name,
                    start_time_unix_nano=start_ns,
                    end_time_unix_nano=end_ns,
                    parent_span_id=parent,
                    attributes=attrs,
                )
            )

        return self._build_payload(spans)

    def export_run_result(
        self,
        run_id: str,
        results: list[dict[str, Any]],
        start_time_ms: float | None = None,
        end_time_ms: float | None = None,
    ) -> OTelExportPayload:
        """Convert evaluation run results into an OTLP payload."""
        trace_id = _generate_trace_id()
        now_ms = time.time() * 1000
        root_start = _ms_to_ns(start_time_ms or now_ms)
        root_end = _ms_to_ns(end_time_ms or now_ms)

        root_span_id = _generate_span_id()
        spans: list[OTelSpan] = [
            OTelSpan(
                trace_id=trace_id,
                span_id=root_span_id,
                name=f"evalgate.run.{run_id}",
                start_time_unix_nano=root_start,
                end_time_unix_nano=root_end,
                attributes=[
                    OTelAttribute(key="evalgate.run_id", value=run_id),
                    OTelAttribute(key="evalgate.service", value=self._service_name),
                ],
            )
        ]

        for r in results:
            span_id = _generate_span_id()
            duration_ms = r.get("duration_ms", 0)
            s_start = _ms_to_ns(r.get("start_time_ms", now_ms))
            s_end = s_start + _ms_to_ns(duration_ms)

            status = 1 if r.get("passed") else 2  # OK or ERROR
            attrs = [
                OTelAttribute(key="evalgate.test_id", value=r.get("test_id", "")),
                OTelAttribute(key="evalgate.test_name", value=r.get("test_name", "")),
                OTelAttribute(key="evalgate.score", value=r.get("score", 0)),
                OTelAttribute(key="evalgate.passed", value=r.get("passed", False)),
            ]
            spans.append(
                OTelSpan(
                    trace_id=trace_id,
                    span_id=span_id,
                    parent_span_id=root_span_id,
                    name=f"evalgate.spec.{r.get('test_name', 'unknown')}",
                    start_time_unix_nano=s_start,
                    end_time_unix_nano=s_end,
                    status_code=status,
                    status_message=r.get("error", ""),
                    attributes=attrs,
                )
            )

        return self._build_payload(spans)

    def _build_payload(self, spans: list[OTelSpan]) -> OTelExportPayload:
        resource = {
            "resource": {
                "attributes": [
                    OTelAttribute(key="service.name", value=self._service_name).to_dict(),
                ],
            },
            "scopeSpans": [
                {
                    "scope": {"name": "evalgate-sdk"},
                    "spans": [s.to_dict() for s in spans],
                }
            ],
        }
        return OTelExportPayload(resource_spans=[resource])

    async def send(self, payload: OTelExportPayload) -> bool:
        """POST the OTLP payload to the collector endpoint.

        Returns False on connection/timeout errors instead of raising.
        """
        url = f"{self._endpoint.rstrip('/')}/v1/traces"
        try:
            async with httpx.AsyncClient(timeout=self._timeout) as client:
                resp = await client.post(
                    url,
                    json=payload.to_dict(),
                    headers={"Content-Type": "application/json", **self._headers},
                )
                return resp.status_code < 400
        except (httpx.ConnectError, httpx.TimeoutException, OSError):
            return False


def create_otel_exporter(options: OTelExporterOptions | None = None) -> OTelExporter:
    """Factory function for OTelExporter."""
    return OTelExporter(options)
